promql/engine.go

```diff
@@ -1802,25 +1824,38 @@ func (ev *evaluator) matrixSelector(node *parser.MatrixSelector) (Matrix, storag
 // values). Any such points falling before mint are discarded; points that fall
 // into the [mint, maxt] range are retained; only points with later timestamps
 // are populated from the iterator.
-func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, maxt int64, out []Point) []Point {
+//** mint, maxt are min timestamp and max timestamp for this slice, in milliseconds
+//** TODO: rename them minTs and maxTs
+func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, maxt int64, extRange bool, out []Point) []Point {
+	//** Drop any samples at the front that we don't need.
+	var dropBefore int
 	if len(out) > 0 && out[len(out)-1].T >= mint {
-		// There is an overlap between previous and current ranges, retain common
+		// There is an overlap between previous and current ranges, so retain common
 		// points. In most such cases:
 		//   (a) the overlap is significantly larger than the eval step; and/or
 		//   (b) the number of samples is relatively small.
 		// so a linear search will be as fast as a binary search.
-		var drop int
-		for drop = 0; out[drop].T < mint; drop++ {
+		//** set dropBefore to index of first data point whose timestamp is inside our range
+		for dropBefore = 0; dropBefore < len(out) && out[dropBefore].T < mint; dropBefore++ {
+		}
+		if extRange {
+			// This is an argument to an extended range function.
+			// Go back one sample if there is one and it's within lookbackDelta of mint.
+			if dropBefore > 0 && out[dropBefore-1].T >= (mint-durationMilliseconds(ev.lookbackDelta)) {
+				dropBefore--
+			}
 		}
-		ev.currentSamples -= drop
-		copy(out, out[drop:])
-		out = out[:len(out)-drop]
 		// Only append points with timestamps after the last timestamp we have.
+		//** reset mint to 1 msec after last timestamp in out
 		mint = out[len(out)-1].T + 1
 	} else {
-		ev.currentSamples -= len(out)
-		out = out[:0]
+		dropBefore = len(out)
 	}
+	//** Drop the samples before dropBefore by copying the remainder over them and
+	//** truncating out.
+	copy(out, out[dropBefore:])
+	out = out[:len(out)-dropBefore]
+	ev.currentSamples -= dropBefore
```

Comment:
I refactored to share the above code. In one case `dropBefore` is right at the end, and hence 0 bytes are copied. That works fine.